### ISSUE TITLE
Implement some way to express "friendship"

### ISSUE THREAD

**Issue by couling** [PyCQA/pylint#4362]
### The problem

Various languages have implemented "friend" access in different ways.  In C++ there's an explicit "[friend](https://en.wikipedia.org/wiki/Friend_function)" modifier, in java the [default scope](https://www.w3schools.com/java/java_modifiers.asp) which allows access to everything from the package.

The need for "friend" functions arises because sometimes a single interface spans over multiple classes and methods.  You can see examples of this in python itself.  Eg: [typing.get_args()](https://github.com/python/cpython/blob/23acadcc1c75eb74b2459304af70d97a35001b34/Lib/typing.py#L1742) accesses `tp.__args__` but this is fine because the interface is the `typing` module not at the class level. 

The problem with linting and`protected-access` warnings is there's is no way to express "friendship" in python leading either excessive comments of the form:

    # pylint disable=protected-access

or it leads to disabling the warning completely for a library or app and reduce the usefulness of this warning.

### Suggested solution

Split `protected-access` (W0212) into two separate warnings, adding a second `protected-module-access` warning so that protected access within the same module does not generate the same warning as protected access to classes outside the module.

This would allow "enabling" something akin to java's default (package) scope by simply disabling `proteced-module-access`.  But by default the change would still warn for all protected access.

**Comment by Pierre-Sassoulas** [comment]
Did you see #4267  ? It will be available in next version.

**Comment by couling** [comment]
@Pierre-Sassoulas as I read it, #4267 is only correcting access within the same class.  It doesn't solve the problem because friendship (where required) is something exclusively about methods outside the class (in other classes or module level methods).

**Comment by NeilGirdhar** [comment]
I just came here looking to see if friend classes had been implemented.  The problem is:
```pythong
class A:
    def _private_method(self): ...

class B:
    def  f(self, a: A):
        a._private_method()
```

I'm not sure the proposed solution is best.  I think the ideal solution is for B to be able to mark itself as a friend of A somehow, which thereafter silences protected access warnings.  Something like:
```python
from typing import Friend

class B(Friend[A]):
    def  f(self, a: A):
        a._private_method()
```
or for the time being
```python
class B:
    # pylint: friend(A)
    def  f(self, a: A):
        a._private_method()
```
What do you think?

**Comment by DanielNoord** [comment]
This seems very difficult to maintain down the line, especially as there is no generally accepted way to do this. It reminds me of the `__implements__` issue.

I think it's better to put this forward to people on `Python Ideas` or `typing-sig` and see if other tools see any merit in supporting this pattern. I don't think we should create our own pattern to express this, but rather follow the general ecosystem after more general discussion.

**Comment by couling** [comment]
> I don't think we should create our own pattern to express this

That's reasonable enough.  It is of course better that linters agree.  

... But alas this is a matter for linters only.  About the only reference I can find to this topic in PEP-8 is

> Public attributes are those that you expect unrelated clients of your class to use, with your commitment to avoid backwards incompatible changes. Non-public attributes are those that are not intended to be used by third parties; you make no guarantees that non-public attributes won’t change or even be removed.

The way I read it linters are already being much stricter than the python standard.  And it would often be acceptable to use protected access from between any two classes in the same distribution (git repo / wheel). It would be pretty difficult for me to go on `Python Ideas` and ask *"hey folks, could we have a feature to allow this already totally allowable concept"*.

So unless there's a place to post this that is a linter-only suggestion for all linters then I guess I'm fresh out of luck.

Thanks very much for your time.

**Comment by NeilGirdhar** [comment]
@couling I suggested this to MyPy, but the suggestion wasn't popular: https://github.com/python/mypy/issues/4740

**Comment by Pierre-Sassoulas** [comment]
The mypy discussion echoes what Daniel said about first making it a thing in typing and the general python ecosystem.

**Comment by tomkcook** [comment]
Something needs to be done here.  Pylint is much too strict.  Consider this case:

```
class Service:
    class Command:
        def __init__(self, service:'Service'):
            reply = queue.Queue()
            self.service = service
        def execute(self):
            self.service._execute(self.reply)
        def get_reply(self):
            return self.reply.get()
            
    def __init__(self):
        self.commands = queue.Queue()
        self.running = True
        self.worker = threading.Thread(target=self.run)
        self.worker.start()

    def run(self):
        while self.running:
            try:
                command = self.commands.get(timeout=1)
                command.execute()
            except queue.Empty:
                pass

    def _execute(self, reply):
        reply.put("Done")

    def Execute(self):
        command = Service.Command()
        self.commands.put(command)
        return command.get_reply()
```

This is a pretty clear implementation of the Command design pattern but there doesn't seem to be any way to make pylint understand that these classes know each other intimately.  Even one the `Command` class is a member of the `Service` class (as above), pylint complains about the `Command` class accessing private members of the `Service` class.

**Comment by Pierre-Sassoulas** [comment]
Could you open a new issue for that @tomkcook, please ?
